diffFiles/src/diff_files.py

```diff
@@ -1,106 +1,130 @@
-            if node1 is None and node2 is None:
-                next_pos = None
-                length = 0
-            elif node2 is not None and (node1 is None or node1.length < node2.length):
-                next_pos = candidate2
-                length = node2.length
-            else:
-                next_pos = candidate1
-                length = cast(Node, node1).length
+@dataclass
+class Matrix:
+    nodes: list[list[Node]]
 
-        return Node(next_pos, length)
+    def get_node(self, pos: Position) -> Node | None:
+        if pos.x < 0 or pos.y < 0:
+            return None
+        return self.nodes[pos.x][pos.y]
 
-    for i1 in range(len(l1)):
-        nodes.append([])
-        for i2 in range(len(l2)):
-            nodes[i1].append(compute_node(i1, i2))
+    @classmethod
+    def make(cls, l1: list[str], l2: list[str]) -> "Matrix":
+        def compute_node(i1: int, i2: int, line1, line2) -> Node:
+            next_pos: Position | None
+            length: int
+
+            if line1 == line2:
+                next_pos = Position(i1 - 1, i2 - 1)
+                length = result.nodes[next_pos.x][next_pos.y].length + 1
+            else:
+                candidate1 = Position(i1, i2 - 1)
+                candidate2 = Position(i1 - 1, i2)
+                node1 = result.get_node(candidate1)
+                node2 = result.get_node(candidate2)
+
+                # find the candidate with the bigger length
+                if node1 is None and node2 is None:
+                    next_pos = None
+                    length = 0
+                elif node2 is not None and (node1 is None or node1.length < node2.length):
+                    next_pos = candidate2
+                    length = node2.length
+                else:
+                    next_pos = candidate1
+                    length = cast(Node, node1).length
+
+            return Node(next_pos, length)
+
+        result = Matrix([])
+        for i1, line1 in enumerate(l1):
+            result.nodes.append([])
+            for i2, line2 in enumerate(l2):
+                result.nodes[i1].append(compute_node(i1, i2, line1, line2))
+        return result
+
+
+def build_subsequence_matrix(l1: list[str], l2: list[str]) -> Matrix:
+    # we use the matrix as a graph
+    # where each node points to one of their neighbors
+    nodes = Matrix.make(l1, l2)
     return nodes
 
 
-def compute_common_indices(nodes: list[list[Node]]) -> list[tuple[int, int]]:
-    pos: tuple[int, int] = (len(nodes) - 1, len(nodes[0]) - 1)
+def compute_common_indices(nodes: Matrix) -> list[Position]:
+    common_indices: list[Position] = []
+    if not nodes.nodes:
+        return common_indices
+
+    pos = Position(len(nodes.nodes) - 1, len(nodes.nodes[0]) - 1)
 
-    def get_node(pos: tuple[int, int]) -> Node:
-        return nodes[pos[0]][pos[1]]
+    def get_node(p: Position) -> Node:
+        result = nodes.get_node(p)
+        if result is None:
+            return Node(Position(0, 0), 0)
```

Comment:
This here is my "tricky" fix for y=-1. It occurs in the "standard-test-case"!

Reply:
I'm a bit puzzled. This is only possible if `len(nodes.nodes[0] - 1) = -1` or `next_pos = (_,-1)`.

The first case occurs if the second file is empty. Second case occurs if I forgot a guard inside `compute_node`. In particular, this would mean that the cast is not current.